### ISSUE TITLE
Recursion

### ISSUE THREAD

**Issue by Yamakaky** [thehydroimpulse/rust-fsm#1]
Hi
Your lib looks quite interesting. However, it uses recursion at each state change, and you can't assume rust will do tail call optimization. Did you consider, instead of recursing, have `switch` to return the new state ? There would be a blocking loop that get the new state and run it, so no stack problem.

**Comment by thehydroimpulse** [comment]
Yeah, this is a pretty old library that I haven't had a chance to update. Having no recursion is definitely necessary!

**Comment by Yamakaky** [comment]
If you don't mind, it's possible that one day I may find the time to update it ^^

**Comment by thehydroimpulse** [comment]
@Yamakaky Oh of course not! Go right ahead.
